File: backend/scoring/executive_function_scoring.py
"""
Executive Function (EF) scoring engine.

STATUS: Organizacionelemnetostareas is fully validated (exact match against the
source SPSS composite across all 1,373 rows). The other six dimensions use a
"best recovered" item set — each carries near-1.0 regression weight and a sharp
drop-off for every other item, but the recovered sum does not exactly reproduce
the original composite for any row (a small residual remains). Evidence strongly
suggests the missing residual comes from item(s) that exist in the original
published ~41-item scale but are not present in this 32-item dataset export.
See EF_scoring_audit_report.md for the full evidence trail.

DO NOT present partial-status scores as equivalent in precision to the exact
dimension, and DO NOT use this engine's output to diagnose ADHD or any other
condition. It measures self-reported executive-function tendencies only.
"""
from dataclasses import dataclass, field
from typing import Dict, List, Optional

# ---------------------------------------------------------------------------
# Item -> dimension formulas (see module docstring + audit report for status)
# ---------------------------------------------------------------------------

DIMENSIONS = {
    'SistemaAtencionalSupervisor': {
        'items': ['EF10', 'EF15', 'EF17', 'EF19', 'EF27', 'EF39'],
        'status': 'partial',           # 'exact' | 'partial'
        'known_full_range': (12, 40),  # observed composite min/max in source data
    },
    'RegulacionDeliberadaEmocion': {
        'items': ['EF4', 'EF34', 'EF40', 'EF42'],
        'status': 'partial',
        'known_full_range': (5, 25),
    },
    'MonitorizacionConscieteResponsabilidades': {
        'items': ['EF2', 'EF8', 'EF9', 'EF18', 'EF38'],
        'status': 'partial',
        'known_full_range': (16, 35),
    },
    'Verificaciondelaconducta': {
        'items': ['EF24', 'EF31', 'EF33', 'EF41'],
        'status': 'partial',
        'known_full_range': (9, 25),
    },
    'Organizacionelemnetostareas': {
        'items': ['EF1', 'EF6', 'EF7', 'EF37'],
        'status': 'exact',
        'known_full_range': (7, 20),
    },
    'Controlinhibitorio': {
        'items': ['EF3', 'EF11', 'EF20', 'EF21', 'EF22', 'EF23'],
        'status': 'partial',
        'known_full_range': (17, 40),
    },
    'tomadedecisiones': {
        'items': ['EF5', 'EF13', 'EF26'],
        'status': 'partial',
        'known_full_range': (7, 25),
    },
}

ALL_ITEMS = sorted(
    {item for d in DIMENSIONS.values() for item in d['items']},
    key=lambda c: int(c[2:])
)

LIKERT_MIN, LIKERT_MAX = 1, 5


class EFValidationError(ValueError):
    pass


@dataclass
class DimensionScore:
    name: str
    status: str            # 'exact' or 'partial'
    items_used: List[str]
    raw_score: int
    mean_score: float
    n_items: int
    theoretical_min: int
    theoretical_max: int
    normalized_0_100: float
    note: Optional[str] = None


@dataclass
class EFScoringResult:
    dimensions: Dict[str, DimensionScore] = field(default_factory=dict)
    overall_score: Optional[float] = None  # intentionally None: no validated overall formula exists yet


def validate_responses(responses: Dict[str, int]) -> None:
    """Raise EFValidationError if responses are missing or out of range.

    Only validates items actually required by at least one dimension formula
    (ALL_ITEMS) -- callers may supply more items, e.g. from the full instrument,
    without triggering an error.
    """
    missing = [item for item in ALL_ITEMS if item not in responses]
    if missing:
        raise EFValidationError(f"Missing required EF item responses: {missing}")

    out_of_range = [
        item for item in ALL_ITEMS
        if not (LIKERT_MIN <= responses[item] <= LIKERT_MAX)
    ]
    if out_of_range:
        raise EFValidationError(
            f"Responses out of range [{LIKERT_MIN}-{LIKERT_MAX}] for items: {out_of_range}"
        )


def score_dimension(name: str, responses: Dict[str, int]) -> DimensionScore:
    spec = DIMENSIONS[name]
    items = spec['items']
    n = len(items)
    raw = sum(responses[i] for i in items)
    theoretical_min, theoretical_max = n * LIKERT_MIN, n * LIKERT_MAX
    normalized = 100 * (raw - theoretical_min) / (theoretical_max - theoretical_min)

    note = None
    if spec['status'] == 'partial':
        note = (
            "Partial recovery: this dimension's original composite could not be "
            "exactly reproduced from the 32 EF items present in this dataset. "
            "The items below carry near-certain (~1.0 regression weight) membership, "
            "but the score may run a few points lower than the original scale intended, "
            "since 1-2 contributing item(s) appear to be absent from this data export."
        )

    return DimensionScore(
        name=name,
        status=spec['status'],
        items_used=items,
        raw_score=raw,
        mean_score=raw / n,
        n_items=n,
        theoretical_min=theoretical_min,
        theoretical_max=theoretical_max,
        normalized_0_100=normalized,
        note=note,
    )


def score_all(responses: Dict[str, int]) -> EFScoringResult:
    """Score every EF dimension from a dict of {item_name: 1-5 response}.

    Does NOT compute an overall EF composite: no validated formula for an
    overall score exists yet (per project scoring-engine requirements, an
    overall score should only be produced once such a formula is established
    and validated).

    This function never produces or implies an ADHD diagnosis.
    """
    validate_responses(responses)
    result = EFScoringResult()
    for name in DIMENSIONS:
        result.dimensions[name] = score_dimension(name, responses)
    return result


if __name__ == '__main__':
    # smoke test with a fabricated all-3 response set
    demo_responses = {item: 3 for item in ALL_ITEMS}
    result = score_all(demo_responses)
    for name, d in result.dimensions.items():
        print(f"{name}: raw={d.raw_score}/{d.theoretical_max} "
              f"mean={d.mean_score:.2f} norm={d.normalized_0_100:.1f} status={d.status}")

File: backend/scoring/executiveFunctionScoring.js
/**
 * Executive Function (EF) scoring — pure arithmetic, no trained model.
 *
 * Unlike ADHD/ASRS detection (RandomForest, needs Python + .pkl files),
 * EF scoring is just summing specific items per dimension. No ML service
 * round-trip needed — this runs directly in the Node backend.
 *
 * STATUS: Organizacionelemnetostareas is exact (validated against the source
 * SPSS composite across all 1,373 rows in the UEF-1 dataset, 0 residual).
 * The other six dimensions use item sets confirmed both by our own
 * regression/brute-force recovery AND by Table 4 of the published paper
 * (Ramos-Galarza et al., 2023, Frontiers in Psychology, UEF-1 scale), but
 * cannot be reproduced exactly from this dataset -- see
 * EF_scoring_audit_report.md for the full evidence trail. Scores for those
 * six will read consistently a little below the published scale's norms.
 *
 * Never use this output to diagnose ADHD or any other condition.
 */

const LIKERT_MIN = 1;
const LIKERT_MAX = 5;

const DIMENSIONS = {
  SistemaAtencionalSupervisor: {
    items: ['EF10', 'EF15', 'EF17', 'EF19', 'EF27', 'EF39'],
    status: 'partial',
  },
  RegulacionDeliberadaEmocion: {
    items: ['EF4', 'EF34', 'EF40', 'EF42'],
    status: 'partial',
  },
  MonitorizacionConscieteResponsabilidades: {
    items: ['EF2', 'EF8', 'EF9', 'EF18', 'EF38'],
    status: 'partial',
  },
  Verificaciondelaconducta: {
    items: ['EF24', 'EF31', 'EF33', 'EF41'],
    status: 'partial',
  },
  Organizacionelemnetostareas: {
    items: ['EF1', 'EF6', 'EF7', 'EF37'],
    status: 'exact',
  },
  Controlinhibitorio: {
    items: ['EF3', 'EF11', 'EF20', 'EF21', 'EF22', 'EF23'],
    status: 'partial',
  },
  tomadedecisiones: {
    items: ['EF5', 'EF13', 'EF26'],
    status: 'partial',
  },
};

const ALL_ITEMS = [...new Set(Object.values(DIMENSIONS).flatMap((d) => d.items))].sort(
  (a, b) => Number(a.slice(2)) - Number(b.slice(2))
);

class EFValidationError extends Error {}

/**
 * @param {Record<string, number>} responses - e.g. { EF1: 3, EF2: 5, ... }
 * @throws {EFValidationError} if any required item is missing or out of range
 */
function validateResponses(responses) {
  const missing = ALL_ITEMS.filter((item) => !(item in responses));
  if (missing.length) {
    throw new EFValidationError(`Missing required EF item responses: ${missing.join(', ')}`);
  }

  const outOfRange = ALL_ITEMS.filter((item) => {
    const v = responses[item];
    return !Number.isInteger(v) || v < LIKERT_MIN || v > LIKERT_MAX;
  });
  if (outOfRange.length) {
    throw new EFValidationError(
      `Responses out of range [${LIKERT_MIN}-${LIKERT_MAX}] for items: ${outOfRange.join(', ')}`
    );
  }
}

function scoreDimension(name, responses) {
  const spec = DIMENSIONS[name];
  const n = spec.items.length;
  const raw = spec.items.reduce((sum, item) => sum + responses[item], 0);
  const theoreticalMin = n * LIKERT_MIN;
  const theoreticalMax = n * LIKERT_MAX;
  const normalized = (100 * (raw - theoreticalMin)) / (theoreticalMax - theoreticalMin);

  return {
    name,
    status: spec.status,
    itemsUsed: spec.items,
    rawScore: raw,
    meanScore: raw / n,
    nItems: n,
    theoreticalMin,
    theoreticalMax,
    normalized0to100: normalized,
    note:
      spec.status === 'partial'
        ? 'Partial recovery: 1-2 contributing item(s) from the original scale are not present ' +
          'in this dataset, so this score runs a few points below what the original instrument ' +
          'would report. Item membership itself is confirmed, not guessed.'
        : null,
  };
}

/**
 * Score all 7 EF dimensions from a flat { EFn: 1-5 } response map.
 * Does NOT compute an overall EF composite -- no validated formula for one exists.
 * Never produces or implies a diagnosis.
 */
function scoreAll(responses) {
  validateResponses(responses);
  const dimensions = {};
  for (const name of Object.keys(DIMENSIONS)) {
    dimensions[name] = scoreDimension(name, responses);
  }
  return { dimensions, overallScore: null };
}

module.exports = {
  DIMENSIONS,
  ALL_ITEMS,
  EFValidationError,
  validateResponses,
  scoreAll,
};
